apiserver/facades/agent/credentialvalidator/credentialvalidator.go

```diff
@@ -101,32 +126,39 @@ func (api *CredentialValidatorAPI) WatchCredential(ctx context.Context, tag para
 	if err != nil {
 		return fail(err)
 	}
-	// Is credential used by the model that has created this backend?
-	modelCredentialTag, exists, err := api.backend.CloudCredentialTag()
+
+	modelInfo, err := api.modelInfoService.GetModelInfo(ctx)
 	if err != nil {
 		return fail(err)
 	}
-	if !exists || credentialTag != modelCredentialTag {
+	exists := modelInfo.CredentialName != ""
+	if !exists {
+		return fail(apiservererrors.ErrPerm)
+	}
+
+	modelCredentialKey := credential.Key{
+		Cloud: modelInfo.Cloud,
+		Owner: modelInfo.CredentialOwner,
+		Name:  modelInfo.CredentialName,
+	}
+	credentialKey := credential.KeyFromTag(credentialTag)
+	if credentialKey != modelCredentialKey {
 		return fail(apiservererrors.ErrPerm)
 	}
 
 	result := params.NotifyWatchResult{}
-	watch, err := api.credentialService.WatchCredential(ctx, credential.KeyFromTag(credentialTag))
+	watcher, err := api.credentialService.WatchCredential(ctx, credentialKey)
 	if errors.Is(err, credentialerrors.NotFound) {
 		err = fmt.Errorf("credential %q %w", credentialTag, errors.NotFound)
 	}
 	if err != nil {
 		result.Error = apiservererrors.ServerError(err)
 		return result, nil
 	}
-	// Consume the initial event. Technically, API calls to Watch
-	// 'transmit' the initial event in the Watch response. But
-	// NotifyWatchers have no state to transmit.
-	if _, ok := <-watch.Changes(); ok {
-		result.NotifyWatcherId = api.resources.Register(watch)
-	} else {
-		watch.Kill()
-		result.Error = apiservererrors.ServerError(watch.Wait())
+	result.NotifyWatcherId, _, err = internal.EnsureRegisterWatcher(ctx, api.watcherRegistry, watcher)
+	if err != nil {
+		api.logger.Infof(ctx, "WatchModelCredential success for notifyWatcherID %q", result.NotifyWatcherId)
```

Comment:
Can delete this log line now.

Reply:
Yup just doing another qa test with the new changes and will delete these, thanks ian! :)

Reply:
QA done and updated the steps, thanks ian!